### PR TITLE
Add import private key progress query

### PR DESCRIPTION
**Pull request by yahtoo**

### CODE REVIEW

File: blockchain/rpc_reactor.go [hunk 1]
```diff
@@ -99,6 +99,7 @@ func (bcr *BlockchainReactor) BuildHandler() {
 
 	m.Handle("/export-private-key", jsonHandler(bcr.walletExportKey))
 	m.Handle("/import-private-key", jsonHandler(bcr.walletImportKey))
+	m.Handle("/import-key-progress", jsonHandler(bcr.importKeyProgress))
```
Comment (Blockmeta) [inline blockchain/rpc_reactor.go]: 这个感觉是一个动词， import

File: blockchain/account/accounts.go [hunk 2]
```diff
@@ -281,7 +276,7 @@ func (m *Manager) createAddress(ctx context.Context, account *Account, change bo
 }
 
 func (m *Manager) createP2PKH(ctx context.Context, account *Account, change bool) (*CtrlProgram, error) {
-	idx := m.nextIndex()
+	idx, _ := m.nextIndex(account)
```
Comment (Paladz) [inline blockchain/account/accounts.go]: 既然这里有了， 还是需要检查error的

File: blockchain/account/accounts.go [hunk 3]
```diff
@@ -308,7 +303,7 @@ func (m *Manager) createP2PKH(ctx context.Context, account *Account, change bool
 }
 
 func (m *Manager) createP2SH(ctx context.Context, account *Account, change bool) (*CtrlProgram, error) {
-	idx := m.nextIndex()
+	idx, _ := m.nextIndex(account)
```
Comment (Paladz) [inline blockchain/account/accounts.go]: 检查error

File: blockchain/wallet/wallet.go [hunk 4]
```diff
@@ -201,7 +266,32 @@ func (w *Wallet) createProgram(account *account.Account, XPub *pseudohsm.XPub, i
 	return nil
 }
 
-//WalletUpdate process every valid block and reverse every invalid block which need to rollback
 func (w *Wallet) rescanBlocks() {
 	w.rescanProgress <- struct{}{}
 }
+
+//GetRescanStatus return key import rescan status
+func (w *Wallet) GetRescanStatus() ([]KeyInfo, error) {
```
Comment (Paladz) [inline blockchain/wallet/wallet.go]: 从w的内存里取数据

File: blockchain/wallet/wallet.go [hunk 5]
```diff
@@ -201,7 +266,32 @@ func (w *Wallet) createProgram(account *account.Account, XPub *pseudohsm.XPub, i
...
+	if err := w.loadKeysInfo(); err != nil {
+		return nil, err
+	}
+	for i, v := range w.keysInfo {
```
Comment (Paladz) [inline blockchain/wallet/wallet.go]: if complete {100%}

File: blockchain/wallet/wallet.go [hunk 6]
```diff
@@ -201,7 +266,32 @@ func (w *Wallet) createProgram(account *account.Account, XPub *pseudohsm.XPub, i
...
+		if v.ImportFlag == true {
+			w.keysInfo[i].Percent = uint8(w.status.Height * 100 / w.chain.Height())
```
Comment (Paladz) [inline blockchain/wallet/wallet.go]: w.chain.Height() ==> walletInfo.Height

File: blockchain/wallet/wallet.go [hunk 7]
```diff
@@ -201,7 +266,32 @@ func (w *Wallet) createProgram(account *account.Account, XPub *pseudohsm.XPub, i
...
+		}
+	}
+	return w.keysInfo, nil
+}
+
+func checkRescanStatus(w *Wallet) {
+	if w.ImportPrivKey {
+		if w.status.Height == w.chain.Height() {
```
Comment (Paladz) [inline blockchain/wallet/wallet.go]: w.chain.Height() ==> walletInfo.Height

File: blockchain/wallet/wallet.go [hunk 8]
```diff
@@ -201,7 +266,32 @@ func (w *Wallet) createProgram(account *account.Account, XPub *pseudohsm.XPub, i
...
+	for i, v := range w.keysInfo {
+		if v.ImportFlag == true {
+			w.keysInfo[i].Percent = uint8(w.status.Height * 100 / w.chain.Height())
+		}
+	}
+	return w.keysInfo, nil
+}
+
+func checkRescanStatus(w *Wallet) {
+	if w.ImportPrivKey {
```
Comment (Paladz) [inline blockchain/wallet/wallet.go]: if !w.ImportPrivKey {return}

File: blockchain/wallet/wallet.go [hunk 9]
```diff
@@ -143,9 +198,10 @@ func (w *Wallet) walletUpdater() {
 func getRescanNotification(w *Wallet) {
 	select {
 	case <-w.rescanProgress:
-		w.status.Height = 1
-		block, _ := w.chain.GetBlockByHeight(w.status.Height)
-		w.status.Hash = block.Hash()
+		w.status.RescanHeight = 1
```
Comment (Paladz) [inline blockchain/wallet/wallet.go]: w.status.RescanHeight = 0

File: blockchain/wallet/wallet.go [hunk 10]
```diff
@@ -21,11 +21,24 @@ import (
 const SINGLE = 1
 
 var walletKey = []byte("walletInfo")
+var privKeyKey = []byte("keysInfo")
 
 //StatusInfo is base valid block info to handle orphan block rollback
 type StatusInfo struct {
-	Height uint64
-	Hash   bc.Hash
+	Height       uint64
```
Comment (Paladz) [inline blockchain/wallet/wallet.go]: Height ==> bestHeight

File: blockchain/wallet/wallet.go [hunk 11]
```diff
@@ -21,11 +21,24 @@ import (
...
+	RescanHeight uint64
+	RescanHash   bc.Hash
+	RescanFlag   bool
+}
+
+//KeyInfo is key import status
+type KeyInfo struct {
+	Alias      string       `json:"alias"`
+	XPub       chainkd.XPub `json:"xpub"`
+	ImportFlag bool         `json:"flag"`
```
Comment (Paladz) [inline blockchain/wallet/wallet.go]: ImportFlag --> delete

File: blockchain/wallet/wallet.go [hunk 12]
```diff
@@ -46,12 +61,17 @@ func NewWallet(walletDB db.DB, account *account.Manager, asset *asset.Registry,
 		AssetReg:       asset,
 		chain:          chain,
 		rescanProgress: make(chan struct{}, 1),
+		ImportPrivKey:  false,
```
Comment (Paladz) [inline blockchain/wallet/wallet.go]: 断电重启状态改变

File: blockchain/wallet/wallet.go [hunk 13]
```diff
@@ -21,11 +21,24 @@ import (
 const SINGLE = 1
 
 var walletKey = []byte("walletInfo")
+var privKeyKey = []byte("keysInfo")
 
 //StatusInfo is base valid block info to handle orphan block rollback
 type StatusInfo struct {
```
Comment (Paladz) [inline blockchain/wallet/wallet.go]: type StatusInfo struct {
 	WorkHeight uint64
 	WorkHash   bc.Hash
 	BestHeight       uint64
 	BestHash         bc.Hash
 	Rescan   bool
 }

File: blockchain/wallet/wallet.go [hunk 14]
```diff
@@ -85,8 +102,33 @@ func (w *Wallet) commitWalletInfo(batch db.Batch) error {
...
+	rawKeysInfo, err := json.Marshal(w.keysInfo)
+	if err != nil {
+		log.WithField("err", err).Error("save wallet info")
+		return err
+	}
+	w.DB.Set(privKeyKey, rawKeysInfo)
+	return nil
+}
+
 func (w *Wallet) attachBlock(block *legacy.Block) error {
```
Comment (Paladz) [inline blockchain/wallet/wallet.go]: in this function, alway compare with WorkHeight && WorkHash

File: blockchain/wallet/wallet.go [hunk 15]
```diff
@@ -114,6 +161,7 @@ func (w *Wallet) detachBlock(block *legacy.Block) error {
 func (w *Wallet) walletUpdater() {
 	for {
 		getRescanNotification(w)
+		checkRescanStatus(w)
 		for !w.chain.InMainChain(w.status.Height, w.status.Hash) {
```
Comment (Paladz) [inline blockchain/wallet/wallet.go]: compare with bestHeight && bestHash

File: blockchain/wallet/wallet.go [hunk 16]
```diff
@@ -127,9 +175,16 @@ func (w *Wallet) walletUpdater() {
 			}
 		}
 
-		block, _ := w.chain.GetBlockByHeight(w.status.Height + 1)
```
Comment (Paladz) [inline blockchain/wallet/wallet.go]: all code above compare with workHeigth && workHash

File: blockchain/account/accounts.go [hunk 17]
```diff
@@ -383,16 +378,27 @@ func (m *Manager) GetCoinbaseControlProgram() ([]byte, error) {
 	return program.ControlProgram, nil
 }
 
-func (m *Manager) nextIndex() uint64 {
+func (m *Manager) nextIndex(account *Account) uint64 {
```
Comment (Paladz) [inline blockchain/account/accounts.go]: 不再是全局的话。。。。。。account结构可以复用么

File: blockchain/wallet/wallet.go [hunk 18]
```diff
@@ -95,27 +141,33 @@ func (w *Wallet) attachBlock(block *legacy.Block) error {
...
 func (w *Wallet) detachBlock(block *legacy.Block) error {
 	storeBatch := w.DB.NewBatch()
 	w.reverseAccountUTXOs(storeBatch, block)
-	w.deleteTransactions(storeBatch, w.status.Height)
+	w.deleteTransactions(storeBatch, w.status.BestHeight)
 
-	w.status.Height = block.Height - 1
-	w.status.Hash = block.PreviousBlockHash
+	w.status.BestHeight = block.Height - 1
+	w.status.BestHash = block.PreviousBlockHash
```
Comment (Paladz) [inline blockchain/wallet/wallet.go]: 少了一个回滚时候handle bestHeight < workHeight的handler

**Pull request merged**